### PR TITLE
initial commit of code changes to add ResultInterface::getNumRows method

### PR DESCRIPTION
**Pull request by sneakyimp**
Also adds the implementation of this method to the various DBMS-specific Result classes. Also added unit test tests/system/database/Live/GetNumRowsTest for this newly added method.

Each pull request should address a single issue and have a meaningful title.

**Description**
Per Lonnie Ezell's [message](https://github.com/codeigniter4/CodeIgniter4/pull/109#issuecomment-752838400) in #109, I have attempted my first CodeIgniter fork/branch/pull request here. It adds a getNumRows method to the ResultInterface and the various DBMS classes that extend it. I've not yet updated the user guide but will try to do so if this looks good.

**Checklist:**
- [X] Securely signed commits
- [X] Component(s) with PHPdocs
- [X] Unit testing, with >80% coverage
- [ ] User guide updated
- [X] Conforms to style guide

### CODE REVIEW

**Comment by sneakyimp** [comment]
OK wow so the automated tests are failing like crazy, but the errors I see don't appear to relate to any files that I've changed. I'm quite new to pull requests and these automated tests. I do know that the one test I added does run without error on my workstation. @lonnieezell perhaps you could help me understand how to pass these tests?

**Comment by sneakyimp** [comment]
I'm going to close this PR and make a cleaner one.

File: system/Database/ResultInterface.php [hunk 1]
```diff
@@ -16,14 +44,15 @@
...
+
 	/**
 	 * Retrieve the results of the query. Typically an array of
 	 * individual data rows, which can be either an 'array', an
 	 * 'object', or a custom class name.
 	 *
 	 * @param string $type The row type. Either 'array', 'object', or a class name to use
 	 *
-	 * @return array
+	 * @return mixed
```
Comment (kenjis) [inline system/Database/ResultInterface.php:55]: Why changed to `mixed`?

File: system/Database/ResultInterface.php [hunk 2]
```diff
@@ -120,8 +149,8 @@ public function getRowObject(int $n = 0);
 	/**
 	 * Assigns an item into a particular column slot.
 	 *
-	 * @param string $key
-	 * @param null   $value
+	 * @param $key
+	 * @param null $value
```
Comment (kenjis) [inline system/Database/ResultInterface.php:153]: `string` is gone.

**Comment by kenjis** [comment]
@sneakyimp 
> PHP Fatal error:  Class CodeIgniter\Test\Mock\MockResult contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (CodeIgniter\Database\ResultInterface::getNumRows) in /home/runner/work/CodeIgniter4/CodeIgniter4/system/Test/Mock/MockResult.php on line 108
https://github.com/codeigniter4/CodeIgniter4/pull/4047/checks?check_run_id=1631979520

You changed `ResultInterface`, so you have to add `getNumRows()` to `MockResult`.
Because `MockResult` extends `BaseResult`, and `BaseResult` implements `ResultInterface`.

File: system/Database/MySQLi/Result.php [hunk 3]
```diff
@@ -58,47 +87,51 @@ public function getFieldNames(): array
 	 */
 	public function getFieldData(): array
 	{
-		static $dataTypes = [
+		static $data_types = [
```
Comment (kenjis) [inline system/Database/MySQLi/Result.php:90]: Why snake_case?

Comment (sneakyimp) [reply system/Database/MySQLi/Result.php:90]: I had mistakenly copied MySQLi/Result.php from a project I've been working on for a few weeks and I guess there were extensive edits to the comments (and this change of var name) in the interim. It became apparent after I submitted this PR that I needed to use more up-to-date files. That is why I closed this PR.

File: system/Database/ResultInterface.php (hunk 1 above)
Comment (sneakyimp) [reply system/Database/ResultInterface.php:55]: As I mentioned elsewhere, this first attempt at a PR had outdated files as its starting point. That is why I closed this request and tried again. Thank you for look at this PR, but I want you to know that I've closed it and it should be ignored. I'm sorry for wasting your time :(

**Pull request closed without merging**